### ISSUE TITLE
Limitations When Updating Kubernetes Versions on VPC Gen 2 Clusters

### ISSUE THREAD

**Issue by bemahone** [IBM-Cloud/terraform-provider-ibm#1978]
### Terraform Version
0.12

### Affected Resource(s)
Please list the resources as a list, for example:
- ibm_container_vpc_cluster

If this issue appears to affect multiple resources, it may be an issue with Terraform's core, so please mention this.

### Important Factoids
Working with IBM Cloud VPC Gen 2 Clusters

### References
Other GitHub issues that appear to be related to updating Kubernetes clusters (NOTE: these are not our colleagues)
- https://github.com/IBM-Cloud/terraform-provider-ibm/issues/1969
- https://github.com/IBM-Cloud/terraform-provider-ibm/issues/1952

### Details
To update a VPC Gen 2 cluster on the IBM Cloud, an `ibmcloud ks worker replace` is issued against a worker.  This removes the original worker and replaces the worker with a new, updated version (if applicable).  

When using the input parameter `update_all_workers` with a VPC Gen 2 cluster, we observe that ALL workers are replaced at the exact same time (in parallel).  This results in ALL workers becoming unavailable at the exact same time, which would violate our SLA with customers.

**In order to effectively service & maintain VPC Gen 2 clusters with Terraform, we need the ability to do the following:**
- Conduct rolling upgrades with no downtime (where 1 worker or a portion of workers are replaced at a time, allowing the cluster to remain operational)
- Support major.minor upgrades via `update_all_workers` (e.g. Kubernetes 1.17 to 1.18) - this exists and works, however it does not currently do this without bringing the entire cluster offline 
- Support worker updates within the existing major.minor level (e.g. 1.17.12_1541 -> 1.17.12_1542) without a change to the master's major.minor being required (e.g. 1.17 to 1.18)
- Support master updates within the existing major.minor level (e.g. 1.17.11_1537 -> 1.17.12_1540)

- The link below, provides examples of how often the clusters are eligible for updates (every 2 weeks).  Many times, our customers stay on one Kubernetes major.minor level for months.  Moving customers to a new major.minor every two weeks to work around this problem would not be feasible.  
  - https://cloud.ibm.com/docs/containers?topic=containers-changelog

- It was suggested to look at the input parameters `workers_info`, `workers_info.id` and `workers_info.version`, however this appears to only work with the `ibm_container_cluster` and not the `ibm_container_vpc_cluster` resource (which is needed for VPC Gen 2 clusters).  

**Please advise if any of our requirements referenced above are on your development roadmap and who we can reach out to in order to advocate for these requirements being added to the features of the `ibm_container_vpc_cluster` resource.**

**Comment by Anil-CM** [comment]
@bemahone we addressed these issues in the PR - https://github.com/IBM-Cloud/terraform-provider-ibm/pull/1989

The PR addresses the following:
1. Serial upgradation of node
2. Separated the dependency of  worker nodes kube version upgradation. - To update only the worker nodes `update_all_workers`. 
3. New parameter - `wait_for_worker_update` is introduced. to avoid more time in updating kube version of worker nodes.
    **Note** : This will cause cluster downtime.

we are coming up with a new resource for kube version upgrade to handle all the requirements.

**Comment by bemahone** [comment]
@Anil-CM - Great news - thank you for the update!  Do you have an estimate as to when all the requirements will be addressed?

**Comment by bemahone** [comment]
Hi @Anil-CM - thanks for the progress on the serial updates to maintain cluster availability...

Here's my most recent findings:
 
Using the following versions:
```Terraform v0.12.28
+ provider.ibm v1.14.0```

... I was able to verify that `update_all_workers` now applies updates to workers one a time.  Interestingly enough while this was happening, the biweekly Kubernetes patch versions were released by the ibm cloud team.  This left me with 4 workers at `1.19.3_1525` and two workers at `1.19.3_1526` (the last two workers updated got the newer version).  This behavior is an expected consequence when not using Terraform as well (we typically avoid staging and production updates on the biweekly patch Mondays, since this can happen).  

When I tried to do another terraform plan/apply (with `update_all_workers` set to true) to get the 4 workers at `1.19.3_1525` to move to `1.19.3_1526`, it did not appear that the `terraform plan` determined any action was needed on the workers.  This was confirmed by the `terraform apply` action as well.  

Is the updating of workers within the same major.minor as the kubernetes master still a work in progress, or was the new expected behavior of `update_all_workers` to move the workers at the `1.19.3_1525` level to the `1.19.3_1526` level?

**Comment by bemahone** [comment]
@Anil-CM - Please see comments in https://github.com/IBM-Cloud/terraform-provider-ibm/issues/1952, where we reference a potential bug with `update_all_workers`.

**Comment by bemahone** [comment]
When setting `update_all_workers` and `wait_for_worker_update ` to `true` and having a set of 3 clusters defined and created by schematics/terraform - we noticed that actions were only taken on the first of the three clusters and the behavior appeared to be a regression from what we previously tested in version 1.14.  

A segment of the terraform containing the count is below:
``` resource "ibm_container_vpc_cluster" "cluster" {
  	count = 3```

**Additional background:**

There was no master update available this Monday 12/7 but there was a worker update available within the same 1.17 cluster level.  Prior tests had taken place with conducting a major.minor upgrade (e.g. 1.17 to 1.18), so this was a slightly different test.    

When we ran the schematics plan/apply - no updates were pushed to the workers or master (nothing was deemed actionable)
...we then toggled off `wait_for_worker_update` setting it to `false` followed by another plan/apply
...we then toggled on `wait_for_worker_update` setting it to `true` followed by another plan/apply
...we observed that was enough to make terraform decide to do something - however, only the first cluster of the set of 3 had any patches applied.   

We also observed that all workers were deleted at the same time, followed by the master.  In the past, when we tested this behavior, it appeared that the workers were deleted and added back one at a time to update.  This behavior seemed different and undesirable from a stability perspective (if running workloads were active, they'd ultimately fail since no workers would be available).  

We will be running some more tests today and report back on additional findings.

**Comment by bemahone** [comment]
Conducted some more tests this week, these are our findings (1.16 provider used within schematics):

**Test 1: Environment with 3 clusters all at the 1.17 IKS level, moved to 1.18 via changing of `kube_version`**
Worked as expected, each master was upgraded first in parallel and 1 worker from each cluster at a time was replaced

**Test 2: Environment with 3 clusters all at the default IKS level (1.18), moved to 1.19 via changing of `kube_version`**
Worked as expected, each master was upgraded first in parallel and 1 worker from each cluster at a time was replaced

**Test 3: Environment with 1 cluster at the default IKS level (1.18), moved to 1.19 via changing of `kube_version`**
Worked as expected, master was upgraded first  and 1 worker was replaced at a time within the cluster 

**Test 4: Environment with 1 cluster at the 1.17 IKS level, moved to 1.18 via changing of `kube_version`**
Worked as expected, master was upgraded first  and 1 worker was replaced at a time within the cluster 

We will see if we can recreate the `wait_for_worker_update` toggle issue during a week where a new patch is available within the same major.minor.  

@Anil-CM - What do you suggest is the best approach to invoke the patching of an IKS cluster during a steady state week, where you need to apply a patch within the same 1.17 family, but there is no need to upgrade from one major.minor to another (e.g. 1.17 -> 1.18)?  Should we get more descriptive with the `kube_version`?  This could get tricky if there isn't a new master level out, but there is a patch for the workers.

**Comment by bemahone** [comment]
@Anil-CM - As of additional tests this evening, where we attempted to update two environments with 3 clusters each, we were unable to invoke any form of steady state maintenance (the generate plan and apply plan actions did nothing to the master and worker node levels).  We tried changing the `kube_version` as well as toggling the `update_all_workers` and `wait_for_worker_update` variables from `true` to `false` back to `true`.  None of these actions had any impact.  It also appeared that if our master was on a 1.18 level (1.18.12_1533) and we specified an override value of `iks_cluster_kube_version` to `1.18.13` - this was not sufficient to make terraform update the master & workers to the latest level (where the latest master level was 1.18.13 at the time of our testing).  

The only way we could get terraform to conduct any updates was jumping from one major.minor to another (e.g. 1.17 -> 1.18  or 1.18 to 1.19 for the `kube_version`).  

As an FYI, we also noticed that there is a new terraform provider running in schematics (1.17) - which was used during the tests this evening.  The change in "toggling" behavior of `wait_for_worker_update` and `update_all_workers` seems to be related to this update, as this behavior was not seen before in 1.16.

**Comment by bemahone** [comment]
@Anil-CM - also observed this evening, that the cluster "upgrade" timed out and the terraform failed to apply mid way through.

```
2020/12/15 23:58:32 Terraform apply | Error: Error waiting for cluster (redacted cluster-id here) worker nodes kube version to be updated: timeout while waiting for state to become 'normal' (last state: 'updating', timeout: 1h0m0s)
2020/12/15 23:58:32 Terraform apply | 
2020/12/15 23:58:32 Terraform apply |   on cluster/cluster.tf line 15, in resource "ibm_container_vpc_cluster" "cluster":
2020/12/15 23:58:32 Terraform apply |   15:  resource "ibm_container_vpc_cluster" "cluster" {
2020/12/15 23:58:32 Terraform apply | 
2020/12/15 23:58:32 Terraform apply | 
2020/12/15 23:58:32 Terraform apply | 
2020/12/15 23:58:32 Terraform apply | Error: Error waiting for cluster (redacted cluster-id here) worker nodes kube version to be updated: timeout while waiting for state to become 'normal' (last state: 'updating', timeout: 1h0m0s)
2020/12/15 23:58:32 Terraform apply | 
2020/12/15 23:58:32 Terraform apply |   on cluster/cluster.tf line 15, in resource "ibm_container_vpc_cluster" "cluster":
2020/12/15 23:58:32 Terraform apply |   15:  resource "ibm_container_vpc_cluster" "cluster" {
2020/12/15 23:58:32 Terraform apply | 
2020/12/15 23:58:32 Terraform apply | 
2020/12/15 23:58:32 Terraform apply | 
2020/12/15 23:58:32 Terraform apply | Error: Error waiting for cluster (redacted cluster-id here) worker nodes kube version to be updated: timeout while waiting for state to become 'normal' (last state: 'updating', timeout: 1h0m0s)
2020/12/15 23:58:32 Terraform apply | 
2020/12/15 23:58:32 Terraform apply |   on cluster/cluster.tf line 15, in resource "ibm_container_vpc_cluster" "cluster":
2020/12/15 23:58:32 Terraform apply |   15:  resource "ibm_container_vpc_cluster" "cluster" {
2020/12/15 23:58:32 Terraform apply | 
2020/12/15 23:58:32 Terraform apply | 
2020/12/15 23:58:32 Terraform APPLY error: Terraform APPLY errorexit status 1
2020/12/15 23:58:32 Could not execute action
```

**Comment by Anil-CM** [comment]
Hi Folks,

At present, terraform provider addresses the following scenarios in k8s cluster update:
Considering the cluster running on version x.y - x& y are major and minor versions
* case-1: Change in kube_version x.y to version x.y1 and update_all_workers is set to default value false. on terraform apply - only kube version change is identified and only the master node's version is updated to x.y1
* case-2: master node version is x.y1, worker nodes are in version x.y and update_all_worker_nodes is set to true. on terraform apply, since no change in the kube_version master update is excluded and goes to update_all_worker_node. if there is version difference between master and any of the worker node, that node will be updated.
* case-3: kube_version of master and worker nodes is x.y1, update_all_workers is set to true. user updates the kube_version from x.y1 to x2.y2 Since there is change in kube_version to x2.y2 , master will be updated to version x2.y2 & as update_all_workers - set to true, update the other worker node's version to x2.y2

These Update scenarios are applicable only for MAJOR and MINOR version updates. Terraform provider is not supported with PATCH updates.

By looking at the time out issue - if more number of worker nodes are there in the cluster, time taken to upgrade the cluster will be more since each worker nodes is repacked serially.
You can add the timeout block in the terraform configuration and configure the update timeout to higher value.

**Comment by bemahone** [comment]
Hi @Anil-CM - Our organization requires us to patch our Kubernetes master and worker nodes when fix-packs/patches become available.  This often occurs once every two weeks.  In particular, we need the ability to apply worker fix-packs within the scope of the same major.minor (e.g. the master `kube_version` of 1.18 or 1.19 in Terraform need not change).  We also need the ability to apply master updates within the same major.minor, when applicable.

The Terraform provider should support fix-pack/patch updates to abide by our organization's security mandates, as these patches address master and worker security vulnerabilities (CVEs).  If left unaddressed, the SecOps team/CIOs office will ultimately shut down the clusters due to their noncompliant posture.  

**Please see the IBM Cloud Kubernetes changelogs (note how there are often many weeks where there are just worker fix packs/patches before another master becomes available and many months between major.minor releases):**
https://cloud.ibm.com/docs/containers?topic=containers-changelog#119_changelog
https://cloud.ibm.com/docs/containers?topic=containers-changelog#118_changelog
https://cloud.ibm.com/docs/containers?topic=containers-changelog#117_changelog

**Comment by bemahone** [comment]
Hi @Anil-CM - thanks for your time today - to recap what we discussed, I am providing a summary below:

We discussed 3 different scenarios that we wanted to verify would be supported by the terraform provider:
 1.  The worker patch level is higher than the master
 2.  The master is already updated and the workers have a different patch level than the master
 3.  The master is already updated and the workers want to apply the same patch level as the master
 
As per our conversation, you mentioned that by removing the verification check to make sure that the workers are being updated to the identical master level, our 3 scenarios would all be supported.  
 
It also appeared that the `patch_version` setting could be loosely defined so that you could ultimately enter whatever you wanted and as long as it changed between a terraform plan/apply, the provider would know to update the workers to the latest available worker version within the context of the master's major.minor level.  

Please let us know when you have a new provider version available that addresses these topics.  

Thanks again!

-Brad & Brian

**Comment by bemahone** [comment]
Hi @Anil-CM - I wanted to share an observation I had last week re: testing `patch_version` on a 1.17 cluster.  I observed that if I didn't provide the master `patch_version` and set the `patch_version` to something else - the terraform/schematics plan would set the value and the apply would do nothing.  However, if I set the `patch_version` to the master level (e.g. `17_1551`) even though the workers were up to date and at the latest available level (`1.17.17_1552`), terraform would replace each worker one at a time.  Just wanted to share this scenario, since a worker replace would not be desired here.

**Comment by Anil-CM** [comment]
> Hi @Anil-CM - I wanted to share an observation I had last week re: testing `patch_version` on a 1.17 cluster. I observed that if I didn't provide the master `patch_version` and set the `patch_version` to something else - the terraform/schematics plan would set the value and the apply would do nothing. However, if I set the `patch_version` to the master level (e.g. `17_1551`) even though the workers were up to date and at the latest available level (`1.17.17_1552`), terraform would replace each worker one at a time. Just wanted to share this scenario, since a worker replace would not be desired here.

Hi @bemahone , this is caused due to comparison of the master's patch_fixpack version with `patch_version`.
The new changes are here in the PR - https://github.com/IBM-Cloud/terraform-provider-ibm/pull/2217

**Comment by bemahone** [comment]
Hi @Anil-CM - thank you for your work to support our feedback!  Do you have an ETA when this fix will be available in a new terraform provider version?

**Comment by hkantare** [comment]
The release v1.21.1 have the fix

**Comment by hkantare** [comment]
can we close this issue

**Comment by bemahone** [comment]
hi @hkantare and @Anil-CM - so far we have had a great experience using this new provider with `patch_version`.  Thank you for your help and incorporating our feedback.  We will be conducting some more end to end patch testing in the next two weeks and I would request that we please leave this open until we have verified all our `patch_version` test scenarios pass.  I will follow-up and close this issue once that is complete.  Thanks!

**Comment by bemahone** [comment]
Thank you for the great progress on the terraform provider.  I am closing this issue and opening another issue #2379 to track the need for additional terraform plan/apply actions to check to make sure the cluster is fully up to date, or if there is outstanding work that needs to be done given a failure/error in the prior terraform plan/apply.
